docs/design-proposals/2021-11-24-graceful-reschedule-tikv-pod.md

```diff
@@ -0,0 +1,137 @@
+# gracefully reschedule tikv pod
+
+TiDB-Operator issue: [#4215](https://github.com/pingcap/tidb-operator/issues/4215)
+
+## Summary
+
+Support gracefully reschedule a tikv pod. The implantation is neccessarrily the combination of three operations:
+
+1. Add a PD evict-leader-scheduler to transfer the leader.
+2. When leader count drop down to zero, delete the pod to let it re-create and reschedule.
+3. Remove the evict-leader-scheduler.
+
+## Motivation
+
+### Goals
+
+- Provide a way for user to gracefully reschedule a tikv pod.
+- The pod must be gracefully stop, to be specify, evict leaders before delete the pod and must work normal when recreate the pod.
+
+### Non-Goals
+
+- Provider a way to make tikv graceful stop when just recive a `SIGTERM` signal.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+suppose `tikv-0` is running at node `node0`,  there are many free resource at other nodes, I would like to make tikv-0 run at other node and shutdown `node0` to reduce the cost. I may apply the flowing operations:
+
+1. run `kubectl cordon node0` to mark `node0` as `unschedulable`.
+2. delete pod `tikv-0` to let it re-create and reschedule to other nodes, to mitigate the impact of unavailability of some region, I will do the flowing operations insted of delete the pod directly:
+   1. add evict-leader-scheduler by `pd-ctl`.
+   2. delete `tikv-0` pod when leader count drop down to 0.
+   3. remove the evict-leader-scheduler by `pd-ctl`
+3. drain `node-0` in the normal way as [safely drain a node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) and shotdown `node0`
+
+### Risks and Mitigations
+
+- PD/TiKV fails to evict all leader and we may just keep checking it.
+
+## Design Details
+
+Support user to add an annotation to tikv pod to trigger an graceful reschedule.
+
+Annotation key `EvictLeaderAnnKey`: `tikv.tidb.pingcap.com/evict-leader`
+
+the value controler the behavier when the leander count drop to zero, valid value is one of:
+
+- `none`: doing nothing
+- `delete-pod`: delete pod and remove the evict-leader scheduler from PD.
+
+
+
+An EvictLeader status willbe added to the `TiKVStatus`:
+
+```go
++type EvictLeaderStatus struct {
++       PodCreateTime metav1.Time `json:"podCreateTime,omitempty"`
++       Value         string      `json:"value,omitempty"`
```

Comment:
do we need to add pod name into this status?

Reply:
https://github.com/pingcap/tidb-operator/pull/4277/commits/edece78c434d7035b303b470ac8e8f3c38f88862
s/*EvictLeader/map[string]*EvictLeader
in tikv status to support multi pod annotated in the same time.